### PR TITLE
add the dataclasses for object metadata and conftest

### LINKED ISSUE #12: make pytest data fixtures for metrics and validation
to return sample inputs that will be available as inputs

### PR DESCRIPTION
**Pull request by alorenzo175**
with examples of what metadata, observations, and forecasts will look like

closes #12

### CODE REVIEW

**Comment by alorenzo175** [comment]
@lboeman @wholmgren what do you think about dataclasses? They are nice because they can be "frozen" and be accessed easily. They will require python 3.7 though.

File: solarforecastarbiter/datamodel.py [hunk 1]
```diff
@@ -0,0 +1,302 @@
...
+        format – these are up to the site owner.
+    """
+    name: str
+    latitude: float
+    longitude: float
+    elevation: float
+    timezone: str
+    network: str = ''
+    well_known_text: str = ''
+    extra_parameters: dict = field(default_factory=dict)
```
Comment (lboeman) [inline solarforecastarbiter/datamodel.py]: If we aren't enforcing the use of json, yaml or csv in extra parameters, perhaps extra_parameters should just be a string?

File: solarforecastarbiter/datamodel.py [hunk 2]
```diff
@@ -0,0 +1,302 @@
...
+    temperature_coefficient : float
+        The temperature coefficient of DC power in units of 1/C.
+        Typically -0.002 to -0.005 per degree C.
+
+    See Also
+    --------
+    FixedTiltModelingParameters
+    SingleAxisModelingParameters
+    """
+    ac_power: float
```
Comment (lboeman) [inline solarforecastarbiter/datamodel.py]: I'm not sure if we need them to be in sync, but ac and dc_power were changed to ac and dc capacity in the proposed dashboard.

File: solarforecastarbiter/datamodel.py [hunk 3]
```diff
@@ -0,0 +1,302 @@
...
+    --------
+    Site
+    SingleAxisModelingParameters
+    """
+    modeling_parameters: SingleAxisModelingParameters = field(
+        default_factory=SingleAxisModelingParameters)
+
+
+@dataclass(frozen=True)
+class FixedTiltPowerPlant(Site):
```
[the remark this answers is not in the document]
Comment (wholmgren) [reply solarforecastarbiter/datamodel.py]: Separating the fixed and tracker at this level does not mirror the API though. I'm not sure what that says about this or the API. Maybe the API needs to look more like the dashboard sketch and like this.

File: solarforecastarbiter/datamodel.py [hunk 4]
```diff
@@ -0,0 +1,302 @@
...
+    --------
+    Site
+    FixedTiltModelingParameters
+    """
+    modeling_parameters: FixedTiltModelingParameters = field(
+        default_factory=FixedTiltModelingParameters)
+
+
+@dataclass(frozen=True)
+class UnitsSetter(GetItem):
```
Comment (wholmgren) [inline solarforecastarbiter/datamodel.py]: A `UnitsSetter` class is a weird thing to inherit. I think it's relatively likely that we'd regret this one. Aren't we supposed to prefer functions to classes? What if I want `UnitsSetter` but not `GetItem`?

File: solarforecastarbiter/datamodel.py [hunk 5]
```diff
@@ -0,0 +1,302 @@
...
+    'dhi': 'W/m^2',
+    'poa_global': 'W/m^2',
+    'relative_humidity': '%',
+    'ac_power': 'MW',
+    'dc_power': 'MW',
+}
+
+
+@dataclass(frozen=True)
+class GetItem(object):
```
[the remark this answers is not in the document]
Comment (alorenzo175) [reply solarforecastarbiter/datamodel.py]: I guess so people could use it like a dict. not necessary though and maybe better without

File: solarforecastarbiter/datamodel.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (alorenzo175) [reply solarforecastarbiter/datamodel.py]: The separate power systems isn't necessary, and maybe it makes more sense to just look at the type of modeling parameters

File: solarforecastarbiter/datamodel.py (hunk 4 above)
Comment (alorenzo175) [reply solarforecastarbiter/datamodel.py]: I don't think we'll regret automatically setting the units from the variable type. now handled by a function

Comment (alorenzo175) [reply solarforecastarbiter/datamodel.py]: and we don't want to modify the dataclass after creation, so should limit any adjustments to __post_init__ and not pass the Forecast object to the function to assign the units

Comment (wholmgren) [reply solarforecastarbiter/datamodel.py]: > I don't think we'll regret automatically setting the units from the variable type.

I agree, my comment was about the inheritance.

> now handled by a function

I like this better.

> and we don't want to modify the dataclass after creation, so should limit any adjustments to **post_init** and not pass the Forecast object to the function to assign the units

makes sense

File: solarforecastarbiter/datamodel.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (wholmgren) [reply solarforecastarbiter/datamodel.py]: I'm not opposed to it if a compelling reason comes along, but I'd rather not start with it.

**Comment by alorenzo175** [comment]
@cwhanse @dplarson @awig Do any of you have reactions to using ``dataclasses`` for the site, forecast, and observation metadata? They provide a nice (almost) immutable container but Python 3.7 is required. I want to make sure 3.7 doesn't hinder development (although a ``conda create -n sfacore python=3.7`` should get an environment going easily).

**Comment by cwhanse** [comment]
We have had issues testing pvlib in py3.7, although off the top of my head, I can't recall what they are.  Something related to numpy functions.

**Comment by wholmgren** [comment]
@cwhanse I think https://github.com/pvlib/pvlib-python/pull/634 solved those issues. I've been using 3.7 in some conda environments for the last few months without issues.

**Comment by cwhanse** [comment]
That was my reaction.  For me `objections is None` from a brief scan of the py 3.7.2 release note.  I'm not currently using py 3.7.

**Comment by alorenzo175** [comment]
@wholmgren Seems like 3.7 is ok. Availability and curtailment were added. anything else?

**Comment by wholmgren** [comment]
merge/rebase on master so your test runs

**Pull request merged**